### PR TITLE
212: More unit tests on RPC

### PR DESCRIPTION
**Pull request by 0ersted**
* refactor crypter

* refactor wallet login

* add unit tests on wallet login

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -72,33 +81,38 @@ void Wallet::Load() {
...
-            }
+    // login session expires
+    if (loginSession) {
+        scheduler_.AddPeriodTask(loginSession, [&]() {
+            threadPool_.Execute([&]() {
+                if (rpcLoggedin_) {
+                    rpcLoggedin_ = false;
+                    spdlog::trace("[Wallet] wallet login session expired!");
+                }
+            });
```
Comment (bassland) [inline src/wallet/wallet.cpp]: I suggest add a task to a thread pool after login

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -149,28 +145,28 @@ class Wallet {
...
     std::pair<uint256, Coin> GetMinerInfo() const;
     void UpdateMinerInfo(uint256 blockHash, const Coin& value);
 
     CKeyID GetLastRedemAddress() const;
-    void SetLastRedemAddress(const CKeyID& lastRedemAddress);
-
     uint256 GetLastRedemHash() const;
-    void SetLastRedemHash(const uint256&);
+
+    void SetLastRedem(const uint256& lastRedemHash, const CKeyID& lastRedemAddress);
```
Comment (qdddddd) [inline src/wallet/wallet.h]:
```suggestion
    void SetLastRedemptInfo(const uint256& lastRedemHash, const CKeyID& lastRedemAddress);
```

File: src/wallet/wallet_store.cpp [hunk 3]
```diff
@@ -234,23 +234,24 @@ std::optional<MasterInfo> WalletStore::GetMasterInfo() {
     return {};
 }
 
-bool WalletStore::StoreLastRedemAddr(const CKeyID& addr) {
+bool WalletStore::StoreLastRedem(const uint256& lastRedemHash, const CKeyID& lastRedemAddress) {
```
Comment (qdddddd) [inline src/wallet/wallet_store.cpp]:
```suggestion
bool WalletStore::StoreLastRedemptInfo(const uint256& lastRedemHash, const CKeyID& lastRedemAddress) {
```

File: src/utils/scheduler.h [hunk 4]
```diff
@@ -46,5 +46,40 @@ class Scheduler {
...
+class Timer {
+public:
+    Timer(uint32_t _duration, std::function<void()>&& f) : duration_(_duration), f_(std::move(f)), stop_(true) {}
+    ~Timer() {
+        Stop();
+    }
+
+    void Reset() {
+        if (duration_ == 0) {
+            return;
```
Comment (BgmLover) [inline src/utils/scheduler.h]: consider adding a log here to indicate the invalid duration

Comment (0ersted) [reply src/utils/scheduler.h]: so I will add a log in the constructor

**Pull request merged**